libminifi/src/core/state/LogMetricsPublisher.cpp

```diff
@@ -0,0 +1,129 @@
+      }
+    }
+    utils::LogUtils::logWithLevel(logger_, log_level_, parent_node.to_pretty_string().c_str());
+  } while (!utils::StoppableThread::waitForStopRequest(logging_interval_));
+}
+
+void LogMetricsPublisher::readLoggingInterval() {
+  gsl_Expects(configuration_);
+  if (auto logging_interval_str = configuration_->get(Configure::nifi_metrics_publisher_log_metrics_logging_interval)) {
+    if (auto logging_interval = minifi::core::TimePeriodValue::fromString(logging_interval_str.value())) {
+      logging_interval_ = logging_interval->getMilliseconds();
+      logger_->log_info("Metric logging interval is set to %lld milliseconds", logging_interval_.count());
+      return;
+    } else {
+      logger_->log_error("Configured logging interval '%s' is invalid!", logging_interval_str.value());
+    }
+  }
+
+  throw Exception(GENERAL_EXCEPTION, "Metrics logging interval not configured for log metrics publisher!");
+}
+
+void LogMetricsPublisher::readLogLevel() {
+  gsl_Expects(configuration_);
+  if (auto log_level_str = configuration_->get(Configure::nifi_metrics_publisher_log_metrics_log_level)) {
+    auto full_caps_log_level_str = utils::StringUtils::toUpper(*log_level_str);
+    log_level_ = utils::LogUtils::LogLevelOption::parse(full_caps_log_level_str.c_str(), utils::LogUtils::LogLevelOption::LOGGING_INFO);
+    logger_->log_info("Metric log level is set to %s", log_level_.toString());
+    return;
+  }
+
+  logger_->log_info("Metric log level is set to INFO by default");
+}
+
+void LogMetricsPublisher::clearMetricNodes() {
+  {
+    std::lock_guard<std::mutex> lock(response_nodes_mutex_);
+    logger_->log_debug("Clearing all metric nodes.");
+    response_nodes_.clear();
+  }
+  if (metrics_logger_thread_) {
+    metrics_logger_thread_->stopAndJoin();
+    metrics_logger_thread_.reset();
+  }
+}
+
+void LogMetricsPublisher::loadMetricNodes() {
+  gsl_Expects(response_node_loader_ && configuration_);
+  auto metric_classes_str = configuration_->get(minifi::Configuration::nifi_metrics_publisher_log_metrics_publisher_metrics);
+  if (!metric_classes_str || metric_classes_str->empty()) {
+    metric_classes_str = configuration_->get(minifi::Configuration::nifi_metrics_publisher_metrics);
+  }
+  bool response_nodes_empty = true;
+  if (metric_classes_str && !metric_classes_str->empty()) {
+    auto metric_classes = utils::StringUtils::split(*metric_classes_str, ",");
+    std::lock_guard<std::mutex> lock(response_nodes_mutex_);
+    for (const std::string& clazz : metric_classes) {
+      auto loaded_response_nodes = response_node_loader_->loadResponseNodes(clazz);
+      if (loaded_response_nodes.empty()) {
+        logger_->log_warn("Metric class '%s' could not be loaded.", clazz);
+        continue;
+      }
+      response_nodes_.insert(response_nodes_.end(), loaded_response_nodes.begin(), loaded_response_nodes.end());
+    }
+    response_nodes_empty = response_nodes_.empty();
+  }
+  if (response_nodes_empty) {
```

Comment:
when can `response_nodes_empty != response_nodes_.empty()` occur?

Reply:
I think it was due to the `response_nodes_` being changed under the lock and didn't want to access it outside of it, but on second thought it's not needed as the lock is only used for synchronizing the logger thread and the changes to the `response_nodes_`, so it does not make any difference. Updated in 70c67450a591c251b152759bab6b2bb8c1422bb5